sdk/appconfiguration/azure-spring-cloud-appconfiguration-config/src/main/java/com/azure/spring/cloud/config/implementation/AppConfigurationPropertySourceLocator.java

```diff
@@ -0,0 +1,291 @@
+
+        StateHolder newState = new StateHolder();
+        newState.setNextForcedRefresh(properties.getRefreshInterval());
+
+        Iterator<ConfigStore> configStoreIterator = configStores.iterator();
+        // Feature Management needs to be set in the last config store.
+        while (configStoreIterator.hasNext()) {
+            ConfigStore configStore = configStoreIterator.next();
+
+            boolean loadNewPropertySources = STARTUP.get() || StateHolder.getLoadState(configStore.getEndpoint());
+
+            if (configStore.isEnabled() && loadNewPropertySources) {
+                // There is only one Feature Set for all AppConfigurationPropertySources
+                FeatureSet featureSet = new FeatureSet();
+
+                List<AppConfigurationReplicaClient> clients = clientFactory
+                    .getAvailableClients(configStore.getEndpoint(), true);
+
+                boolean generatedPropertySources = false;
+
+                List<AppConfigurationPropertySource> sourceList = new ArrayList<>();
+                boolean reloadFailed = false;
+
+                for (AppConfigurationReplicaClient client : clients) {
+                    sourceList = new ArrayList<>();
+
+                    if (!STARTUP.get() && reloadFailed
+                        && !AppConfigurationRefreshUtil.checkStoreAfterRefreshFailed(configStore.getFeatureFlags(), client,
+                            clientFactory)) {
+                        // This store doesn't have any changes where to refresh store did. Skipping Checking next.
+                        continue;
+                    }
+
+                    // Reverse in order to add Profile specific properties earlier, and last profile
+                    // comes first
+                    try {
+                        sourceList
+                            .addAll(create(client, configStore, profiles, !configStoreIterator.hasNext(), featureSet));
+
+                        LOGGER.debug("PropertySource context.");
+
+                        // Setting new ETag values for Watch
+                        List<ConfigurationSetting> watchKeysSettings = new ArrayList<>();
+                        List<ConfigurationSetting> watchKeysFeatures = new ArrayList<>();
+
+                        for (AppConfigurationStoreTrigger trigger : configStore.getMonitoring().getTriggers()) {
+                            ConfigurationSetting watchKey = client.getWatchKey(trigger.getKey(), trigger.getLabel());
+                            if (watchKey != null) {
+                                watchKeysSettings.add(watchKey);
+                            } else {
+                                watchKeysSettings
+                                    .add(new ConfigurationSetting().setKey(trigger.getKey())
+                                        .setLabel(trigger.getLabel()));
+                            }
+                        }
```

Comment:
Seems like we can extract this section to a method?

Reply:
Which returns a `watchKeysSettings`

Reply:
We can, it will have to be a different method than the watch key one for feature flags. This one goes off a list of given watch keys, the other is looking at all feature flags and does a list.